From 8be1d49961088d0993d9f742642708ba5b4a7632 Mon Sep 17 00:00:00 2001
From: Holger Weiss <holger@zedat.fu-berlin.de>
Date: Wed, 29 Jun 2016 22:32:59 +0200
Subject: [PATCH] mod_mam_mnesia: Force garbage collection

The VM fails to collect the garbage generated during MAM lookups
automatically, so mod_mam_mnesia's memory usage easily goes up to
several gigabytes if we don't force garbage collection.

Applied in 16.06+1.

---
 src/mod_mam_mnesia.erl | 15 +++++++++------
 1 file changed, 9 insertions(+), 6 deletions(-)

diff --git a/src/mod_mam_mnesia.erl b/src/mod_mam_mnesia.erl
index 2463690..be14d0f 100644
--- a/src/mod_mam_mnesia.erl
+++ b/src/mod_mam_mnesia.erl
@@ -138,12 +138,15 @@ select(_LServer, JidRequestor,
     SortedMsgs = lists:keysort(#archive_msg.timestamp, Msgs),
     {FilteredMsgs, IsComplete} = filter_by_rsm(SortedMsgs, RSM),
     Count = length(Msgs),
-    {lists:map(
-       fun(Msg) ->
-	       {Msg#archive_msg.id,
-		jlib:binary_to_integer(Msg#archive_msg.id),
-		mod_mam:msg_to_el(Msg, MsgType, JidRequestor, JidArchive)}
-       end, FilteredMsgs), IsComplete, Count}.
+    Result = {lists:map(
+		fun(Msg) ->
+			{Msg#archive_msg.id,
+			 jlib:binary_to_integer(Msg#archive_msg.id),
+			 mod_mam:msg_to_el(Msg, MsgType, JidRequestor,
+					   JidArchive)}
+		end, FilteredMsgs), IsComplete, Count},
+    erlang:garbage_collect(),
+    Result.
 
 %%%===================================================================
 %%% Internal functions
